### PR TITLE
[FLINK-3888] Allow custom convergence criterion in delta iterations

### PR DESCRIPTION
**Pull request by vasia**
As discussed in the Jira issue, this PR contains the following changes:
1. use `TaskConfig.setConvergenceCriterion()` to set the custom, user-defined convergence criterion (like in the case of bulk iteration)
2. add a new method `TaskConfig.setDefaultConvergeCriterion()` to handle the default empty workset convergence
​3. check both criteria in `IterationSynchronizationSinkTask.checkForConvergence()​`
4. expose the custom convergence criterion in `DeltaIteration`
It also contains some minor cleanup and corresponding changes in the `CollectionExecutor`.
The iteration docs already state that custom convergence is possible, so no update needed there ;)

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/iterative/aggregators/AggregatorConvergenceITCase.java [hunk 1]
```diff
@@ -137,42 +138,51 @@ public void testConnectedComponentsWithParametrizableConvergence() {
...
+			DataSet<Tuple2<Long, Long>> updatedComponentId =
+					verticesWithNewComponents.join(iteration.getSolutionSet()).where(0).equalTo(0)
+							.flatMap(new MinimumIdFilter(UPDATED_ELEMENTS));
+
+			List<Tuple2<Long, Long>> result = iteration.closeWith(updatedComponentId, updatedComponentId).collect();
+			Collections.sort(result, new JavaProgramTestBase.TupleComparator<Tuple2<Long, Long>>());
+
+			assertEquals(expectedResult, result);
+		}
+		catch (Exception e) {
```
Comment (greghogan) [inline flink-tests/src/test/java/org/apache/flink/test/iterative/aggregators/AggregatorConvergenceITCase.java]: Catching and re-throwing exceptions looks to be unnecessary.

File: flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java [hunk 2]
```diff
@@ -1513,14 +1513,21 @@ private void finalizeWorksetIteration(IterationDescriptor descr) {
...
+				throw new CompilerException("Error: Aggregator convergence criterion set, but aggregator is null.");
+			}
+
+			syncConfig.setConvergenceCriterion(convAggName, convCriterion);
 		}
 		
 		headConfig.addIterationAggregator(WorksetEmptyConvergenceCriterion.AGGREGATOR_NAME, new LongSumAggregator());
 		syncConfig.addIterationAggregator(WorksetEmptyConvergenceCriterion.AGGREGATOR_NAME, new LongSumAggregator());
-		syncConfig.setConvergenceCriterion(WorksetEmptyConvergenceCriterion.AGGREGATOR_NAME, new WorksetEmptyConvergenceCriterion());
+		syncConfig.setDefaultConvergenceCriterion(WorksetEmptyConvergenceCriterion.AGGREGATOR_NAME, new WorksetEmptyConvergenceCriterion());
```
Comment (vasia) [inline flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java]: Sure that's possible, but each iteration will have its own TaskConfig.

**Comment by vasia** [comment]
Thank you for the review @greghogan! I have addressed your comments.

**Comment by greghogan** [comment]
+1 to merge

**Comment by vasia** [comment]
Thank you Greg! Merging.

**Pull request closed without merging**